server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -170,14 +153,14 @@ private QueryShardContext(int shardId,
     }
 
     private void reset() {
-        allowUnmappedFields = indexSettings.isDefaultAllowUnmappedFields();
+        this.searchFields.setAllowUnmappedFields(indexSettings.isDefaultAllowUnmappedFields());
         this.lookup = null;
         this.namedQueries.clear();
         this.nestedScope = new NestedScope();
     }
 
     public Similarity getSearchSimilarity() {
-        return similarityService != null ? similarityService.similarity(mapperService::fieldType) : null;
+        return similarityService != null ? similarityService.similarity(searchFields::fieldType) : null;
```

Comment:
here I left the function, I thought it would be a shame to pass through the whole SearchFields where only the field type lookup is needed. Open to ideas.